File: Data Structures & Algorithms/reverse-a-linked-list/submission-15.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
     ListNode current = head;
     ListNode tmp;
     ListNode previous = null;

     while (current != null) {
        tmp = current.next;
        // update the links
        current.next = previous;

        // counter update rule 
        previous = current;
        current = tmp;
     }

     return previous;   
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-15.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return

        slow = head
        fast = head.next

        # 1. find middle
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # 2. reverse second half
        second = slow.next
        slow.next = None
        prev = None

        while second:
            tmp = second.next
            
            second.next = prev
            prev = second
            second = tmp

        # 3. merge two halves
        first = head
        second = prev

        while second:
            tmp1 = first.next
            tmp2 = second.next

            first.next = second
            second.next = tmp1

            first = tmp1
            second = tmp2